pkg/cpio/newc/newc.go

```diff
@@ -117,6 +117,9 @@ func (w *writer) WriteRecord(f cpio.Record) error {
 
 	buf := &bytes.Buffer{}
 	hdr := headerFromInfo(f.Info)
+	if f.ReadCloser == nil {
```

Comment:
This turned out to be needed for correctness in producing newc records. How is size set to non-zero? It's set to non-zero by a stat of several things, including directories. Should producers of cpio records set it to zero? Well, for some formats, maybe, for others maybe not. Maybe there is some format that would be fine with a non-zero directory size and even use it as a hint when making a directory (AOS/VS had that; you could set the maximum # elements in a directory). 

So the question is, is the file size == 0 for newc correctly enforced in newc record generators or in newc? I kind of went back and forth on this and I'm still not sure. But it's not really a result of an incorrect record by any means. It's an ambiguity in cpio in my opinion. There's no requirement I can see, anywhere, that says every single format writes directory records with size 0. 

But I don't know, maybe it's best just to accept that we will only ever do newc. 

This actually fixes a real problem, which could be created by someone passing a perfectly legitimate record to newc. If we are passed a directory record with a non-zero size, and we naively just write the non-zero size out as part of the record, then many cpios implementing newc will get an error on read -- even though, arguably, they should know better: directory size has no real meaning in newc as far as I can tell, and ought to just be ignored in 'i' mode.

The newc format kind of requires that directory size be zero. But would all formats require that?
 That's not clear. 
Not totally disagreeing with you here, but not really agreeing either.